rfcs/0024-hard-forks.md

```diff
@@ -0,0 +1,455 @@
+    abort the upgrade entirely, releasing new clients that extend the current
+    era up to the next date.
+5. One week from era start:
+  * Mainnet client builds that support the new era are released and all nodes
+    are expected to upgrade. If no further intervention occurs the transition
+    will proceed as planned.
+  * No further changes are allowed. If a bug that should block era transition is
+    found we delay the era start by another week, releasing new clients, and
+    return to step 4.
+6. Era start:
+  * All changes are active on the mainnet and the burn-in testnet is retired.
+    Pop the confetti.
+
+For emergency fixes the procedure is much shorter. Anything that threatens
+security of funds, network liveness, or consistency counts as an emergency. We
+could also do emergency forks to revert thefts or loss of funds due to error,
+as in Ethereum's [DAO fork](https://eips.ethereum.org/EIPS/eip-779), but whether
+and when to do that is outside the scope of this RFC.
+
+Timeline:
+1. The underlying issue is discovered and documented. The Coda team decides an
+   emergency hard fork is necessary.
+2. We schedule the new era, for some time in the near future e.g. 48 hours from
+   now. We release new client versions that halt after the next scheduled era
+   starts. This reduces the chance a client will be stuck on a dead network. If
+   the fix is very simple we can skip this.
+3. We write and test the fix, then release new builds, with the new era
+   programmed in.
+4. The new era begins.
+
+If an emergency fix is done in the six weeks prior to a scheduled upgrade, the
+upgrade will be delayed such that the new era starts seven weeks after the
+hotfix era does - with the six week process above beginning one week after the
+hotfix era starts to give us some buffer.
+
+### Technical considerations
+
+#### Distinguishing blocks from different eras
+
+We extend each block with an `era_id` field, which is a 32 bit integer. Each ID
+is used at most once, and corresponds to a specific set of features. They are
+not necessarily sequential, and are distinct from era numbers, which are. If we
+change what features are in a particular era, we use a new ID. Each client
+executable includes the set of global slot interval to era id mappings, and
+checks that every block it accepts has the appropriate era id for its slot. The
+blockchain SNARK certifies that the current and prior era IDs are correct. A
+block with a global slot that has no corresponding era ID is invalid.
+
+#### Feature flags
+
+In order to develop different features in parallel and avoid having painful long
+lived branches, we'll adopt feature flags for blocks and transactions. This is
+essential for disabling individual features that aren't ready yet without
+breaking other code or requiring extensive repo surgery.
+
+Feature flags will be determined statically, and the changes to types will be
```

Comment:
This is a nice way to do feature flags, we should consider immediately adopting something like this to land small PRs for incomplete features